rust/yarp-sys/build.rs

```diff
@@ -0,0 +1,168 @@
+        .allowlist_type("yp_pack_size")
+        // TODO: Commenting this because I can't figure out how to get bindgen to generate the
+        // inner-unions nicely. Hand-rolling this in src/lib.rs for now.
+        // .allowlist_type("yp_string_t")
+        .blocklist_type(r#"^yp_string_t\S*"#)
+        .allowlist_type("yp_string_list_t")
+        .allowlist_type("yp_token_type_t")
+        // Enums
+        .rustified_non_exhaustive_enum("yp_comment_type_t")
+        .rustified_non_exhaustive_enum("yp_context_t")
+        .rustified_non_exhaustive_enum("yp_heredoc_indent_t")
+        .rustified_non_exhaustive_enum("yp_heredoc_quote_t")
+        .rustified_non_exhaustive_enum("yp_lex_mode_t")
+        .rustified_non_exhaustive_enum("yp_lex_state_t")
+        .rustified_non_exhaustive_enum("yp_node_type")
+        .rustified_non_exhaustive_enum("yp_pack_encoding")
+        .rustified_non_exhaustive_enum("yp_pack_endian")
+        .rustified_non_exhaustive_enum("yp_pack_length_type")
+        .rustified_non_exhaustive_enum("yp_pack_result")
+        .rustified_non_exhaustive_enum("yp_pack_signed")
+        .rustified_non_exhaustive_enum("yp_pack_size")
+        .rustified_non_exhaustive_enum("yp_pack_type")
+        .rustified_non_exhaustive_enum("yp_pack_variant")
+        .rustified_non_exhaustive_enum("yp_token_type")
+        .rustified_non_exhaustive_enum("yp_unescape_type_t")
+        // Functions
+        .allowlist_function("yp_buffer_init")
+        .allowlist_function("yp_buffer_free")
+        .allowlist_function("yp_node_destroy")
+        .allowlist_function("yp_list_empty_p")
+        .allowlist_function("yp_list_free")
+        .allowlist_function("yp_list_init")
+        .allowlist_function("yp_node_memsize")
+        .allowlist_function("yp_pack_parse")
+        .allowlist_function("yp_parse")
+        .allowlist_function("yp_parse_serialize")
+        .allowlist_function("yp_parser_free")
+        .allowlist_function("yp_parser_init")
+        .allowlist_function("yp_parser_register_encoding_changed_callback")
+        .allowlist_function("yp_parser_register_encoding_decode_callback")
+        .allowlist_function("yp_prettyprint")
+        .allowlist_function("yp_regexp_named_capture_group_names")
+        .allowlist_function("yp_serialize")
+        .allowlist_function("yp_size_to_native")
+        .allowlist_function("yp_string_free")
+        .allowlist_function("yp_string_length")
+        .allowlist_function("yp_string_source")
+        .allowlist_function("yp_string_list_init")
+        .allowlist_function("yp_string_list_free")
+        .allowlist_function("yp_token_type_to_str")
+        .allowlist_function("yp_unescape_calculate_difference")
+        .allowlist_function("yp_unescape_manipulate_string")
+        .allowlist_function("yp_version")
+        // Vars
+        .allowlist_var(r#"^yp_encoding\S+"#)
+        .generate()
+        .expect("Unable to generate yarp bindings")
```

Comment:
Seeing this pretty long list of things, including things which are not really public API (like enum), I wonder if it wouldn't be much simpler and safer to use serialization for Rust instead: https://github.com/ruby/yarp/discussions/1068#discussioncomment-6708266

@kddnewton My impression is the C API is not stable yet and it's not public API either, except for the functions marked with `YP_EXPORTED_FUNCTION`. As such it feels wrong that Rust would depend on so many implementation details that are subject to change.

Reply:
You can see how little API surface is needed when using serialization at https://github.com/ruby/yarp/pull/1112

Reply:
I would very much prefer to use the C API, the same one MRI uses, instead of serialization.

Reply:
I would not use this library if it relied on serialization.

Reply:
Why not? Could you explain the reasons?

Basically because you depend on internals here you must expect on every release something could break and you'd need to adjust your Rust code using it, even after 1.0, because it's not depending just on the public API (currently just `YP_EXPORTED_FUNCTION`) but on internals too.
And it would be an overhead for YARP maintenance.

For instance, why does this expose `yp_lex_mode_t` to Rust? That seems strictly internal.

I suppose there should be a clear definition of what part of YARP's C API is considered public/fairly stable and what part is internal (and you are on your own if you depend on that).
`YP_EXPORTED_FUNCTION` functions is a start but is too restrictive.
We'd need at least node structs (if there is a field rename there it would break Rust code using it, but seems unavoidable anyway).
Things like parser errors/warnings/comments we'd probably need some functions exposing it cleanly rather than directly the `yp_list_t`.
Basically the fewer structs/enum exposed the better maintenance-wise.

Reply:
I'm fine with no relying on the serialization API. I do think we can investigate whether templating the structs would be more beneficial to the ergonomics than using bindgen, but we'll need bindgen either way for the functions.

Reply:
> Why not? Could you explain the reasons?

At this point, my only interest is for building a Ruby LSP in Rust. I've used `tree-sitter` and [lib-ruby-parser](https://github.com/lib-ruby-parser/lib-ruby-parser) for some prototypes; both have their pros and cons, but neither are ideal for a Ruby LSP, IMHO. Because requirements of server-side LSP can require some heavy lifting, I really care about eeking out every bit of performance I can. While `yarp` looks amazing for lots of things, I'm still concerned it might not perform in this setting as well as `tree-sitter`, given `tree-sitter`'s explicit features for working with text. I only imagine relying on serialization/deserialization for all of this text would only mean a performance hit, and I'd rather take the complexity and churn if that means I have a Ruby LSP implementation that's not slow.

> Basically because you depend on internals here you must expect on every release something could break and you'd need to adjust your Rust code using it, even after 1.0, because it's not depending just on the public API (currently just YP_EXPORTED_FUNCTION) but on internals too.
And it would be an overhead for YARP maintenance.

I initially pinged @kddnewton in #1068 to see if there was interest; they said "yes", so I made the PR. If there wasn't interest in the code, I'd totally understand that (specifically due to the maintenance), and would happily make these same exact bindings available outside the `ruby` organization (i.e. I'd release them on my own). Like @lopopolo has alluded to, I think that bindings at this level have the potential to be really helpful for others, and if they're part of the `ruby` org, I imagine they'll only receive more attention (which means higher quality) than if they were provided by a lone developer such as myself.

If I've added bindings for anything that's not public, that was an accident, and those things should be removed--specifically due to the argument you're giving. These bindings should only be for the public API, and should cover the least surface area possible while allowing to use that API. If what I've put up is not that, then we should change it.

> Basically the fewer structs/enums exposed the better maintenance-wise.

💯 

> I'm no Rust expert but I would expect it's so much nicer to use if using serialization and template-generated Rust classes (one per node type), because then you don't need to wrap everything in unsafe { }, including node field accesses, Rust owns the memory and knows how to deal with it efficiently, etc.

As I think you could probably tell, at this point there's no code in here that adds any functionality on top of the C API--it effectively just adds the ability to call that API. As I think you're alluding to, to be really useful though, one would want to make another Rust library that builds on these bindings. When building one of these wrapper libs in Rust, the expectation is that you'll always be using `unsafe` and that's fine/expected/the-only-way-to-do-this (I mean, Rust itself uses `unsafe` in its own codebase...). (Sorry if I'm pointing out obvious things--just trying to be thorough) IMO, since those next-level-up wrapper libraries can be opinionated, it could make sense to release just these bindings, then wait for some time (months?) to see what surfaces in the community before introducing "Ruby's official Rust yarp lib" (<- the higher level crate that would abstract away all the `unsafe` and make it nice and ergonomic/Rusty). That'd give time for `yarp` itself to stabilize _and_ give time for these bindings to stabilize, all while providing others (myself included) to use them for their needs. (...and now that I think about it, personally, I don't think I'd need a full-blown wrapper of these bindings--I'd just want to be able to call `parse()`, navigate through the parse tree, and match against `node` variants in that parse tree...)

As far as "Rust owns the memory"... while this is certainly contentious, I really actually like the idea of borrowing/sharing any memory I can from yarp because hopefully that means less copying/cloning/whatever, and hopefully better performance.

Reply:
Thank you for the detailed reply.

> I'd rather take the complexity and churn if that means I have a Ruby LSP implementation that's not slow.

I think it is a rather small overhead and I would bet this overhead is not what makes a Ruby LSP slow or fast.
In fact in it was found in  https://github.com/ruby/yarp/pull/1149#issuecomment-1646219164 that the difference is small and also in https://github.com/ruby/yarp/discussions/836#discussioncomment-5960781 (although the Ruby API via C extension copies nodes, which I think is so much simpler and necessary if one must be able to create nodes in Ruby; Also consider this is deserializing in pure Ruby code, it would be so much faster in Rust).
IIRC when I measured parsing+serializing and deserializing in Java for TruffleRuby I found that parsing is about 10x slower than deserializing, IOW deserializing is like a 10% (+ a bit more for serializing, but that should be faster than deserializing as it doesn't need to allocate anything except the buffer) overhead.

> If I've added bindings for anything that's not public, that was an accident, and those things should be removed--specifically due to the argument you're giving. These bindings should only be for the public API, and should cover the least surface area possible while allowing to use that API. If what I've put up is not that, then we should change it.

Alright, then I think probably only parsing (`yp_parse` and its dependencies, including all the node structs) should be exposed to Rust, maybe lexing too if that's useful. @kddnewton mentioned we should probably have a public header and internal header to clarify what is what. That is for sure a useful thing to do and define.

> As far as "Rust owns the memory"... while this is certainly contentious, I really actually like the idea of borrowing/sharing any memory I can from yarp because hopefully that means less copying/cloning/whatever, and hopefully better performance.

I can see why for "read-only" workloads it feels tempting to just borrow the memory.
But if it's to gain 10% is it really worth all the noise in caller code using `unsafe {}` etc?
Also not being able to use node flags easily and all other nice things defined on node classes in Ruby and Java, but not in C because well C doesn't have classes.
I'm genuinely interested how much the performance difference would be and how much easier to use it would be.
I think it'd be worth to try both.

BTW do you really have some code/prototype using this? That could be useful to assess "what needs to be exposed" and most of the rest probably just doesn't need to be exposed.

Reply:
@eregon you are over indexing on `unsafe` blocks. This is what it means to call extern C APIs in Rust. You should imagine the unsafe blocks around every API call to yarp in MRI. The two usages are equivalent.

As @turboladen explained, it is very unlikely that normal Rust code would directly consume `yarp-sys` and thus would not need to use unsafe blocks. Idiomatic wrappers around foreign code include both the raw bindings (the sys crate) and a higher level safe wrapper which is idiomatic, ergonomic rust with a safe public API.

This crate's sole purpose is to make the public C API available to Rust code. You can think of the Rust code generated by bindgen to be a set of extern declarations in a C or C++ header. No more, no less.

Rust has native, easy interop with C libraries. It would be an odd restriction to force Rust programs to go through the serialization APIs if the yarp maintainers do not have that expectation for C code that consumes this library.

Reply:
> Alright, then I think probably only parsing (yp_parse and its dependencies, including all the node structs) should be exposed to Rust, maybe lexing too if that's useful.

I think that, along with getting diagnostic info ("why did parsing fail?") would work for me. For my part, I don't think I have a use for lexing at this point.

Reply:
> BTW do you really have some code/prototype using this? 

I have 3 repos of LSP prototypes, going back 2 years or so, but they're all a bit of a mess. :D Everything is pretty straightforward until I want to account for multiple definitions of something (class, method, etc) that exist in different files/workspaces/dependencies. But like I mentioned above, my bare-bones needs are:

- a `parse()` function (obviously :) ), which returns some collection of nodes (AST, CST, whatever) and diagnostic info (so I can report back to the client where code in a file is invalid)
- node info (type, location of the whole node) for each node in the collection
- the ability to navigate (walk, traverse, whatever) the collection would be nice. Ideally, it'd be really easy to look up nodes that are relevant to the node where the cursor is: if the cursor is inside an instance method on a class, I want to know all the things that are in scope. Surely there are lots of ways to solve this, but I find myself wanting to implement that functionality ("give me all the nodes in scope X") on the collection of nodes that I've parsed. Probably out of `yarp`'s scope, but thought I'd throw that out there. :)

I'm guessing it'd also be out of `yarp`'s scope, but I do like how in tree-sitter, you can "edit" a parse tree in place: https://docs.rs/tree-sitter/0.20.10/tree_sitter/struct.Tree.html#method.edit. I've assumed, of course, that this functionality helps with efficiency, but never ran any benchmarks.